File: java/ods/Treap.java
package ods;

import java.util.Comparator;
import java.util.Random;


public class Treap<T> extends
		BinarySearchTree<Treap.Node<T>, T> implements SSet<T> {
	/**
	 * A random number source
	 */
	Random rand;

	protected static class Node<T> extends BinarySearchTree.BSTNode<Node<T>,T> {
		int p;
	}
	
	public Treap(Comparator<T> c) {
		super(new Node<T>(), c);
		rand = new Random();
	}

	public Treap() {
		this(new DefaultComparator<T>());	
	}
	
	public boolean add(T x) {
		Node<T> u = newNode();
		u.x = x;
		u.p = rand.nextInt();
		if (super.add(u)) {
			bubbleUp(u);
			return true;
		}
		return false;
	}

	protected void bubbleUp(Node<T> u) {
		while (u.parent != nil && u.parent.p > u.p) {
			if (u.parent.right == u) {
				rotateLeft(u.parent);
			} else {
				rotateRight(u.parent);
			}
		}
		if (u.parent == nil) {
			r = u;
		}
	}

	public boolean remove(T x) {
		Node<T> u = findLast(x);
		if (u != nil && c.compare(u.x, x) == 0) {
			trickleDown(u);
			splice(u);
			return true;
		}
		return false;
	}

	/**
	 * Do rotations to make u a leaf
	 */
	protected void trickleDown(Node<T> u) {
		while (u.left != nil || u.right != nil) {
			if (u.left == nil) {
				rotateLeft(u);
			} else if (u.right == nil) {
				rotateRight(u);
			} else if (u.left.p < u.right.p) {
				rotateRight(u);
			} else {
				rotateLeft(u);
			}
			if (r == u) {
				r = u.parent;
			}
		}
	}
}

File: python/ods/treap.py
"""An implementation of Treaps/Cartesian trees

This is an implementation of the data structure called a Treap by Aragon
and Seidel:

C. R. Aragon and R. Seidel. Randomized Search Trees. In Algorithmica, 
   Vol. 16, Number 4/5, pp. 464-497, 1996.

Pretty-much the same structure was discovered earlier by Vuillemin:

J. Vuillemin. A unifying look at data structures. 
   Communications of the ACM, 23(4), 229-239, 1980. 
"""
import random

from binarysearchtree import BinarySearchTree

class Treap(BinarySearchTree):
    class Node(BinarySearchTree.Node):
        def __init__(self, x):
            super(Treap.Node, self).__init__(x)
            self.p = random.random()
            
        def __str__(self):
            return "[%r,%f]" % (self.x, self.p)
            
    def __init__(self, iterable=[]):
        super(Treap, self).__init__(iterable)
    
    def _new_node(self, x):
        return Treap.Node(x)
        
    def add(self, x):
        u = self._new_node(x)
        if self.add_node(u):
            self.bubble_up(u)
            return True
        return False
            
    def bubble_up(self, u):
        while u != self.r and u.parent.p > u.p:
            if u.parent.right == u:
                self.rotate_left(u.parent)
            else:
                self.rotate_right(u.parent)

        if u.parent == self.nil:
            self.r = u
            
    def remove(self, x):
        u = self._find_last(x)
        if u is not None and u.x == x:
            self.trickle_down(u)
            self.splice(u)
            return True
        return False
        
    def trickle_down(self, u):
        while u.left is not None or u.right is not None:
            if u.left is None:
                self.rotate_left(u)
            elif u.right is None:
                self.rotate_right(u)
            elif u.left.p < u.right.p:
                self.rotate_right(u)
            else:
                self.rotate_left(u)
            if self.r == u:
                self.r = u.parent
        